apps/remote_control/lib/lexical/remote_control/search/store/backends/cub/state.ex

```diff
@@ -0,0 +1,242 @@
+  def find_by_subject(%__MODULE__{} = state, type, subtype, subject) do
+    versions = Versions.current()
+
+    CubDB.with_snapshot(state.cub_pid, fn snap ->
+      snap
+      |> CubDB.Snapshot.select(min_key: min_subject_key(), max_key: max_subject_key())
+      |> Stream.flat_map(fn
+        {subject_key(subject: db_subject), entries} ->
+          Enum.filter(entries, fn entry ->
+            correct_version_range?(entry, versions) and
+              subject_matches?(db_subject, subject) and
+              type_and_subtype_match?(entry, type, subtype)
+          end)
+      end)
+      |> Enum.to_list()
+    end)
+  end
+
+  def find_by_references(%__MODULE__{} = state, type, subtype, references) do
+    versions = Versions.current()
+
+    all_ref_keys = Enum.map(references, fn ref -> ref_key(id: ref) end)
+
+    state.cub_pid
+    |> CubDB.get_multi(all_ref_keys)
+    |> Stream.filter(fn {_, entry} ->
+      correct_version_range?(entry, versions) and
+        type_and_subtype_match?(entry, type, subtype)
+    end)
+    |> Enum.map(fn {_, entry} -> entry end)
+  end
+
+  def select_all(%__MODULE__{} = state) do
+    min_key = min_ref_key()
+    max_key = max_ref_key()
+
+    state.cub_pid
+    |> CubDB.select(min_key: min_key, max_key: max_key)
+    |> Enum.map(fn {_, value} -> value end)
+  end
+
+  defp new_subject_key(%Entry{} = entry) do
+    subject_key(subject: to_subject(entry.subject), type: entry.type, subtype: entry.subtype)
+  end
+
+  defp entries_to_keys_and_values(entries) do
+    by_subject = Enum.group_by(entries, &new_subject_key(&1), & &1)
+
+    by_path =
+      Enum.group_by(
+        entries,
+        &path_key(path: &1.path),
+        &{new_subject_key(&1), ref_key(id: &1.ref)}
+      )
+
+    by_ref = Enum.map(entries, &{ref_key(id: &1.ref), &1})
+
+    Enum.concat([by_subject, by_path, by_ref])
+  end
+
+  defp subject_matches?(db_subject, query_subject) do
+    do_subject_matches?(db_subject, to_subject(query_subject))
+  end
+
+  defp do_subject_matches?(_, :_), do: true
+
+  defp do_subject_matches?(same_subject, same_subject) when is_binary(same_subject) do
+    true
+  end
+
+  defp do_subject_matches?(_, _), do: false
+
+  defp correct_version_range?(%Entry{} = entry, versions) do
+    entry.elixir_version == versions.elixir and entry.erlang_version == versions.erlang
+  end
+
+  defp type_and_subtype_match?(%Entry{type: type}, type, :_) do
+    true
+  end
+
+  defp type_and_subtype_match?(%Entry{subtype: subtype}, :_, subtype) do
+    true
+  end
+
+  defp type_and_subtype_match?(%Entry{type: type, subtype: subtype}, type, subtype) do
+    true
+  end
+
+  defp type_and_subtype_match?(%Entry{}, _type, _subtype) do
+    false
+  end
+
+  defp min_subject_key do
+    subject_key(
+      type: min_for_type(:atom),
+      subtype: min_for_type(:atom),
+      subject: min_for_type(:binary)
+    )
+  end
+
+  defp max_subject_key do
+    subject_key(type: max_for_type(:atom), subtype: max_for_type(:atom))
+  end
```

Comment:
Does this work? I thought we needed to wrap binaries in `{binary}` so that the min key could be `{}` and the max `{"", ""}`.

Reply:
Oh, right, it works because having a max key early in the key means later ones don't matter. It will matter, though, if we end up doing narrower selects specifying the type/subtype.